### PR TITLE
SI-6574 Support @tailrec for extension methods.

### PR DESCRIPTION
**Pull request by retronym**
Currently, when the body of an extension method is transplanted
to the companion object, recursive calls point back to the original
instance method. That changes during erasure, but this is too late
for tail call analysis/elimination.

This commit eagerly updates the recursive calls to point to the
extension method in the companion. It also removes the @tailrec
annotation from the original method.

Review by @adriaanm

### CODE REVIEW

**Comment by retronym** [comment]
Hmmm.

```
[quick.repl] Compiling 48 files to /localhome/jenkins/a/workspace/pr-rangepos-per-commit/build/quick/classes/repl
[quick.repl] error: Unknown type: <notype>, <notype> [class scala.reflect.internal.Types$NoType$, class scala.reflect.internal.Types$NoType$] TypeRef? false
[quick.repl] error: scala.reflect.internal.FatalError: Unknown type: <notype>, <notype> [class scala.reflect.internal.Types$NoType$, class scala.reflect.internal.Types$NoType$] TypeRef? false
[quick.repl]    at scala.reflect.internal.SymbolTable.abort(SymbolTable.scala:53)
[quick.repl]    at scala.tools.nsc.Global.abort(Global.scala:256)
[quick.repl]    at scala.tools.nsc.backend.icode.TypeKinds$class.toTypeKind(TypeKinds.scala:401)
[quick.repl]    at scala.tools.nsc.backend.icode.ICodes.toTypeKind(ICodes.scala:18)
[quick.repl]    at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadApply1$1(GenICode.scala:607)
[quick.repl]    at scala.tools.nsc.backend.icode.GenICode$ICodePhase.scala$tools$nsc$backend$icode$GenICode$ICodePhase$$genLoad(GenICode.scala:634)
[quick.repl]    at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadQualifier(GenICode.scala:1080)
[quick.repl]    at scala.tools.nsc.backend.icode.GenICode$ICodePhase.genLoadApply6$1(GenICode.scala:773)
```

Shutting this one down for now.

File: src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala [hunk 1]
```diff
@@ -264,4 +266,33 @@ abstract class ExtensionMethods extends Transform with TypingTransformers {
           stat
       }
   }
+
+  final class SubstututeRecursion(origMeth: Symbol, extensionMeth: Symbol,
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala:270]: I like the u-u-u sound, but u-i-u is generally more accepted. Anyway, can be fixed some other day.

File: src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala [hunk 2]
```diff
@@ -264,4 +266,33 @@ abstract class ExtensionMethods extends Transform with TypingTransformers {
...
+      // SI-6574 Rewrite recursive calls against the extension method so they can
+      //         be tail call optimized later. The tailcalls phases comes before
+      //         erasure, which performs this translation more generally at all call
+      //         sites.
+      //
+      //         // Source
+      //         class C[C] { def meth[M](a: A) = { { <expr>: C[C'] }.meth[M'] } }
+      //
+      //         // Translation
+      //         class C[C] { def meth[M](a: A) = { { <expr>: C[C'] }.meth[M'](a1) } }
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/ExtensionMethods.scala:282]: Whule we're nutpuckung, use T1...Tn for the type params rather than reusung C?

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**